app/models/carto_uploader.rb

```diff
@@ -33,29 +49,74 @@ def upload(data)
       end
       keys = keys.join(",")
       vals = vals.join(",")
-      allVals.push("(#{vals})");  #For Mass Insertion mode only.
+      arrVals.push("(#{vals})");
       #--------------------------------
-
-      #Option 1: Individual SQL Insertions
+      
+      #If we have enough classifications, do a batch Insert.
       #--------------------------------
-      if @mode != CARTODB_INSERT_MODE_MASS
-        sqlQuery = "INSERT INTO #{CARTODB_TABLE} (#{keys}) VALUES (#{vals}) "
-        uri = URI(CARTODB_URI_GET.gsub(/__SQLQUERY__/, URI.escape(sqlQuery)))
-        res = Net::HTTP.get(uri)
-        puts res
+      if (arrVals.length >= CLASSIFICATIONS_PER_BATCH)
+        expectedInsertions += arrVals.length
+        successfulInsertions += batchInsert(keys, arrVals)
+        arrVals = []
       end
       #--------------------------------
     end
 
-    #Option 2: Mass SQL Insertions
+    #Do a final batch Insert for any stragglers
+    #--------------------------------
+    if (arrVals.length > 0)
+      expectedInsertions += arrVals.length
+      successfulInsertions += batchInsert(keys, arrVals)
+      arrVals = []
+    end
     #--------------------------------
-    allVals = allVals.join(",")
-    if @mode == CARTODB_INSERT_MODE_MASS
+    
+    #Final sanity check?
+    #--------------------------------
+    puts "Inserted #{successfulInsertions} Classifications out of #{expectedInsertions}"
+    if (expectedInsertions != successfulInsertions)
+      #TODO
```

Comment:
WAIT, CRAP.

I just realised that the `if (expectedInsertions != successfulInsertions)` check should happen after **every batchInsert()**.

Otherwise, imagine what happens if BatchA = success, BatchB = failure, BatchC = success. If that happens, BatchB wouldn't be able to be uploaded again, because as far as the system sees, BatchC already has the 'latest' Classification ID.

Also, edge case: what happens if the last item in BatchB has the same Classification ID as the first item in BatchC? Unlikely to happen in practice, but something to keep in mind if we ever have to manually "force redo" an upload from, say, BatchC after BatchB has successfully gone through.